enterprise/backend/src/metabase_enterprise/advanced_permissions/driver/impersonation.clj

```diff
@@ -19,27 +20,23 @@
 
 (set! *warn-on-reflection* true)
 
-(defn- enforce-impersonation?
-  "Takes the permission set for each group a user is in, and an impersonation policy, and determines whether the policy
-  should be enforced. This is done by checking whether the union of permissions in all *other* groups provides full
-  data access to the database. If so, we don't enforce the policy, because theo ther groups' permissions supercede it."
-  [group-id->perms-set {db-id :db_id}]
-  (let [perms-set (apply set/union (vals group-id->perms-set))]
-    (not (perms/set-has-full-permissions? perms-set (perms/all-schemas-path db-id)))))
-
-(defn- enforced-impersonations
+(defn- enforce-impersonations?
   "Given a list of Connection Impersonation policies and a list of permission group IDs that the current user is in,
-  filter the policies to only include ones that should be enforced for the current user. An impersonation policy is
-  not enforced if the user is in a different permission group that grants full access to the database."
-  [impersonations group-ids]
+  returns a Boolean indicating whether the policies should be enforced. They are not enforced if any of the *other*
+  groups the user is in provide unrestricted data access to the DB."
+  [db-or-id impersonations group-ids]
   (let [non-impersonated-group-ids (set/difference (set group-ids)
                                                    (set (map :group_id impersonations)))
         perms                      (when (seq non-impersonated-group-ids)
-                                     (t2/select Permissions {:where [:in :group_id non-impersonated-group-ids]}))
-        group-id->perms-set        (-> (group-by :group_id perms)
-                                       (update-vals (fn [perms] (into #{} (map :object) perms))))]
-    (filter (partial enforce-impersonation? group-id->perms-set)
-            impersonations)))
+                                     (t2/select :model/DataPermissions {:where
```

Comment:
nit, maybe arguable: could be clearer to do `(t2/select-fn-set :perm_value ...)` to select the permission set directly, instead of mapping it into a set in the next step?